packages/backend/src/apps/m365-excel/common/workbook-session/redis.ts

```diff
@@ -0,0 +1,113 @@
+import Redlock from 'redlock'
+
+import { type M365TenantInfo } from '@/config/app-env-vars/m365'
+import {
+  makeRedisAppDataKey,
+  redisAppDataClient,
+} from '@/helpers/redis-app-data'
+
+import { APP_KEY } from '../constants'
+
+//
+// In general, we store file-specific data in our app data redis DB, with keys
+// following the naming convention below:
+//
+// <M365 Tenant ID>:<File ID>:<...description of data>
+//
+// This function helps to make sure our redis operations keep to this key naming
+// convention.
+//
+function makeRedisKeyPrefix(tenant: M365TenantInfo, fileId: string): string {
+  return makeRedisAppDataKey(APP_KEY, `${tenant.id}:${fileId}:`)
+}
+
+//
+// Locking stuff
+//
+
+// This is the lock used as a write lock for session IDs
+//
+// Note that it allows for retrying a little bit, instead of failing
+// immediately if it fails to acquire the lock. Although retries will increase
+// our worker latency, it reduces full action re-queues (since the default is to
+// throw a retry-step error), and also makes front-end UX better (retry-step
+// errors surface as real errors in the pipe editor, instead of being retried).
+//
+// Session manipulation should be pretty fast plus this should be quite rare, so
+// we _should_ be OK to take the occasional latency hit.
+const redlock = new Redlock([redisAppDataClient], {
+  retryCount: 2,
+  retryDelay: 500,
+  retryJitter: 100,
+})
+
+// Redlock has a timeout to guard against lock holder crashes.
+//
+// This timeout is arbitrarily set to 150s - a little bit more default timeout
+// for 1 HTTP request. The idea is that a lock holder should release the lock if
+// its HTTP request has failed; otherwise, it should be able to extend the lock
+// before continuing.
+const DEAD_LOCK_HOLDER_PREVENTION_TIMEOUT_MS = 150 * 1000
+
+export async function runWithLock<T>(
+  tenant: M365TenantInfo,
+  fileId: string,
+  callback: Parameters<typeof redlock.using<T>>[2],
+): Promise<T> {
+  const lockKey = `${makeRedisKeyPrefix(tenant, fileId)}session:lock`
+
+  return await redlock.using(
+    [lockKey],
+    DEAD_LOCK_HOLDER_PREVENTION_TIMEOUT_MS,
+    callback,
+  )
+}
+
+//
+// Session management stuff
+//
+
+// Session ID redis key expiry
+//
+// Arbitrarily set to 1 hour. See giant comment in index.ts for more details.
+const SESSION_ID_EXPIRY_SECONDS = 60 * 60
+
+export async function getSessionIdFromRedis(
+  tenant: M365TenantInfo,
+  fileId: string,
+): Promise<string | null> {
+  const key = `${makeRedisKeyPrefix(tenant, fileId)}session:id`
+
+  const [[hgetErr, sessionId], [expireErr]] = await redisAppDataClient
```

Comment:
```suggestion
  const [[getErr, sessionId], [expireErr]] = await redisAppDataClient
```

does await throw the error or return it as a response here?

Reply:
await returns the error in the response (e.g. to test, change the expire command to an invalid time string - e.g. `10o`.)

According to this old ioredis [discussion](https://github.com/redis/ioredis/issues/301#issuecomment-218641172), It looks like ioredis has "split" error handling:
- outright syntax errors -> throw error
- command execution error -> return error

One of the contributions mentioned wanting to change this in ioredis v3, but as far as I can tell this double-error handling status quo has remained.